File: main/src/main/java/net/fwitz/math/main/binary/escape/newton/NewtonsMethodPlot.java
package net.fwitz.math.main.binary.escape.newton;

import net.fwitz.math.binary.complex.Complex;
import net.fwitz.math.fractal.escape.EscapeFunction;
import net.fwitz.math.plot.binary.escape.color.NewtonsMethodDarkZero;
import net.fwitz.math.plot.binary.escape.EscapeTimePlot;

import java.util.function.Function;

import static net.fwitz.math.fractal.escape.newton.NewtonsMethod.newtonsMethod;

public class NewtonsMethodPlot {
    private NewtonsMethodPlot() {
    }

    public static void renderEscape(String name, EscapeFunction escapeFn) {
        EscapeTimePlot.complex(name)
                .computeFn(escapeFn)
                .domainX(-3.5, 3.5)
                .domainY(-2.5, 2.5)
                .size(700, 500)
                .colorFn(new NewtonsMethodDarkZero())
                .render();
    }

    public static void render(String name, Function<Complex, Complex> f) {
        renderEscape(name, newtonsMethod(f));
    }

    public static void render(String name, Function<Complex, Complex> f, Function<Complex, Complex> df) {
        renderEscape(name, newtonsMethod(f, df));
    }

    public static void zPowMinus1(double power) {
        render("z^" + power + " - 1 (Newton's Method)",
                z -> z.pow(power).minus(1),
                z -> z.pow(power - 1).times(power));
    }

    public static void zPowMinus1(Complex power) {
        render("z^" + power + " - 1",
                z -> z.pow(power).minus(1),
                z -> z.pow(power.minus(1)).times(power));
    }
}

File: main/src/main/kotlin/net/fwitz/math/main/binary/escape/newton/NewtonsMethodPlot.kt
package net.fwitz.math.main.binary.escape.newton

import net.fwitz.math.binary.complex.Complex
import net.fwitz.math.fractal.escape.EscapeFunction
import net.fwitz.math.fractal.escape.newton.NewtonsMethod
import net.fwitz.math.fractal.escape.newton.NewtonsMethod.newtonsMethod
import net.fwitz.math.plot.binary.escape.EscapeTimePlot
import net.fwitz.math.plot.binary.escape.color.NewtonsMethodDarkZero
import java.util.function.Function

object NewtonsMethodPlot {
    fun renderEscape(name: String, escapeFn: EscapeFunction<Complex>) {
        EscapeTimePlot.complex(name)
            .computeFn(escapeFn)
            .domainX(-3.5, 3.5)
            .domainY(-2.5, 2.5)
            .size(700, 500)
            .colorFn(NewtonsMethodDarkZero())
            .render()
    }

    fun render(name: String, f: (Complex) -> Complex) {
        renderEscape(name, newtonsMethod(f))
    }

    fun render(name: String, f: (Complex) -> Complex, df: (Complex) -> Complex) {
        renderEscape(name, newtonsMethod(f, df))
    }

    fun zPowMinus1(power: Double) = render(
        name = "z^$power - 1 (Newton's Method)",
        f = { z -> z.pow(power) - 1 },
        df = { z -> z.pow(power - 1) * power }
    )

    fun zPowMinus1(power: Complex) = render(
        name = "z^$power - 1",
        f = { z -> z.pow(power) - 1 },
        df = { z -> z.pow(power - 1) * power }
    )
}